go/ai/evaluator.go

```diff
@@ -0,0 +1,240 @@
+// Copyright 2024 Google LLC
+// SPDX-License-Identifier: Apache-2.0
+
+package ai
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/firebase/genkit/go/core"
+	"github.com/firebase/genkit/go/core/logger"
+	"github.com/firebase/genkit/go/core/tracing"
+	"github.com/firebase/genkit/go/internal/atype"
+	"github.com/firebase/genkit/go/internal/registry"
+	"github.com/google/uuid"
+	"go.opentelemetry.io/otel/trace"
+)
+
+// Evaluator represents a evaluator action.
+type Evaluator interface {
+	// Name returns the name of the evaluator.
+	Name() string
+	// Evaluates a dataset.
+	Evaluate(ctx context.Context, req *EvaluatorRequest) (*EvaluatorResponse, error)
+}
+
+type (
+	evaluatorActionDef core.ActionDef[*EvaluatorRequest, *EvaluatorResponse, struct{}]
+
+	evaluatorAction = core.ActionDef[*EvaluatorRequest, *EvaluatorResponse, struct{}]
+)
+
+// Example is a single example that requires evaluation
+type Example struct {
+	TestCaseId string   `json:"testCaseId,omitempty"`
+	Input      any      `json:"input"`
+	Output     any      `json:"output,omitempty"`
+	Context    []any    `json:"context,omitempty"`
+	Reference  any      `json:"reference,omitempty"`
+	TraceIds   []string `json:"traceIds,omitempty"`
+}
+
+// Dataset is a collection of [Example]
+type Dataset = []Example
+
+// EvaluatorRequest is the data we pass to evaluate a dataset.
+// The Options field is specific to the actual evaluator implementation.
+type EvaluatorRequest struct {
+	Dataset      *Dataset `json:"dataset"`
+	EvaluationId string   `json:"evalRunId"`
+	Options      any      `json:"options,omitempty"`
+}
+
+// ScoreStatus is an enum used to indicate if a Score has passed or failed. This
+// drives additional features in tooling / the Dev UI.
+type ScoreStatus int
+
+const (
+	ScoreStatusUnknown ScoreStatus = iota
+	ScoreStatusFail
+	ScoreStatusPass
+)
+
+var statusName = map[ScoreStatus]string{
+	ScoreStatusUnknown: "unknown",
+	ScoreStatusFail:    "fail",
+	ScoreStatusPass:    "pass",
+}
+
+func (ss ScoreStatus) String() string {
+	return statusName[ss]
+}
+
+// Score is the evaluation score that represents the result of an evaluator.
+// This struct includes information such as the score (numeric, string or other
+// types), the reasoning provided for this score (if any), the score status (if
+// any) and other details.
+type Score struct {
+	Id      string         `json:"id,omitempty"`
+	Score   any            `json:"score,omitempty"`
+	Status  string         `json:"status,omitempty" jsonschema:"enum=unknown,enum=fail,enum=pass"`
```

Comment:
Is it okay to use jsonschema to validate the values for this enum?

Reply:
You'll be the first... we don't really rely on this feature so I can't say whether it works well or not. Generally we do manual checking of all inputs.

Reply:
Ack. This is on the response path and unlikely to be prone to user error. Will leave this as it is now.